### PR TITLE
avoid temporary vectors

### PR DESCRIPTION
**Pull request by matthieugomez**
Simple fixes to avoid temporary vectors + adapt the norm computation to sparse matrices. This makes the algorithm an order of magnitude faster for sparse matrices.

### CODE REVIEW

**Comment by matthieugomez** [comment]
I have just updated the pull request — code is now simpler.

File: src/trust_region.jl [hunk 1]
```diff
@@ -20,39 +20,42 @@ macro trustregiontrace(stepnorm)
...
+        p_i = J \ r # Gauss-Newton step
     catch e
         if isa(e, Base.LinAlg.LAPACKException) || isa(e, Base.LinAlg.SingularException)
             # If Jacobian is singular, compute a least-squares solution to J*x+r=0
             U, S, V = svd(full(J)) # Convert to full matrix because sparse SVD not implemented as of Julia 0.3
             k = sum(S .> eps())
-            mrinv = V*diagm([1./S[1:k]; zeros(length(S)-k)])*U' # Moore-Penrose generalized inverse of J
-            p_i = -mrinv*r
+            mrinv = V * diagm([1./S[1:k]; zeros(length(S)-k)]) * U' # Moore-Penrose generalized inverse of J
+            p_i = mrinv * r
```
Comment (spencerlyon2) [inline src/trust_region.jl]: It looks like you are removing negative signs. Is this intentional?

File: test/runtests.jl [hunk 2]
```diff
@@ -13,7 +13,13 @@ tests = ["2by2.jl",
 
 println("Running tests:")
 
-for t in tests
-    println(" * $(t)")
-    include(t)
+for test in tests
+    try
```
Comment (spencerlyon2) [inline test/runtests.jl]: Why do we now need this try/catch?

Are some tests now failing that were previously passing?

**Comment by spencerlyon2** [comment]
I added a column to the minpack test printouts to track total time.

Current master:

```
Running tests:
 * 2by2.jl
 * singular.jl
 * finite_difference.jl
 * minpack.jl
Function                           Dim    NFEV    NJEV   Final inf-norm     total time
--------------------------------------------------------------------------------------
Rosenbrock                           2      30      17     0.000000e+00   6.768056e-03
Rosenbrock-AD                        2      30      17     0.000000e+00   7.704676e-02
Powell singular                      4      17      17     2.945101e-09   1.575939e-02
Powell singular-AD                   4      17      17     2.945101e-09   1.175167e-02
Powell badly scaled                  2      15      14     7.051582e-12   1.241603e-02
Powell badly scaled-AD               2      15      14     7.051582e-12   1.250186e-02
Wood                                 4      19      15     1.152565e-10   2.783019e-02
Wood-AD                              4      19      15     1.152336e-10   2.151693e-02
Helical Valley                       3      10       9     1.110223e-14   1.861635e-02
Helical Valley-AD                    3      10       9     1.110223e-14   3.491276e-02
Watson                               6      20      16     2.830843e-13   5.294593e-02
Watson-AD                            6      20      16     2.886788e-13   5.551840e-02
Watson                               9      17      17     1.618251e-12   5.651870e-04
Watson-AD                            9      17      17     1.599735e-12   1.516064e-02
Chebyquad                            5       8       5     9.143862e-09   2.150277e-02
Chebyquad-AD                         5       8       5     9.143862e-09   1.753459e-02
Chebyquad                            6      13       8     7.320533e-16   1.960040e-04
Chebyquad-AD                         6      13       8     7.320533e-16   5.408860e-04
Chebyquad                            7      10       6     6.329061e-10   9.363300e-05
Chebyquad-AD                         7      10       6     6.329061e-10   3.246310e-04
Chebyquad                            9      14       8     1.649028e-14   1.913390e-04
Chebyquad-AD                         9      14       8     1.649028e-14   7.500110e-04
Brown almost-linear                 10       5       5     3.382330e-09   7.364024e-02
Brown almost-linear-AD              10       5       5     3.382331e-09   6.419869e-02
Brown almost-linear                 30       4       4     8.883264e-09   2.161200e-04
Brown almost-linear-AD              30       4       4     8.883264e-09   2.186250e-04
Brown almost-linear                 40       4       4     1.570838e-09   2.197280e-04
Brown almost-linear-AD              40       4       4     1.570838e-09   2.563530e-04
Discrete boundary value             10       4       4     2.636780e-16   2.391723e-02
Discrete boundary value-AD          10       4       4     2.636780e-16   1.548664e-02
Discrete integral equation           1       4       4     8.540391e-14   2.810263e-02
Discrete integral equation-AD        1       4       4     8.540391e-14   2.462884e-02
Discrete integral equation          10       4       4     2.720046e-15   1.337460e-04
Discrete integral equation-AD       10       4       4     2.720046e-15   5.850890e-04
Trigonometric                       10      14       8     1.204317e-10   1.977184e-02
Trigonometric-AD                    10      14       8     1.204299e-10   1.555886e-02
Variably dimensioned                10      15      15     1.323497e-12   2.322180e-02
Variably dimensioned-AD             10      15      15     1.323497e-12   1.335670e-02
Broyden tridiagonal                 10       5       5     7.548402e-10   1.616605e-02
Broyden tridiagonal-AD              10       5       5     7.548402e-10   1.324463e-02
Broyden banded                      10       6       6     9.359466e-09   2.467619e-02
Broyden banded-AD                   10       6       6     9.359466e-09   1.426321e-02
 * iface.jl
 * already_converged.jl
 * autodiff.jl
 * josephy.jl
 * difficult_mcp.jl
 * sparse.jl
```

With this PR:

```
Running tests:
	PASSED: 2by2.jl
	PASSED: singular.jl
	PASSED: finite_difference.jl
Function                           Dim    NFEV    NJEV   Final inf-norm     total time
--------------------------------------------------------------------------------------
Rosenbrock                           2      30      17     0.000000e+00   6.388667e-03
Rosenbrock-AD                        2      30      17     0.000000e+00   8.225130e-02
Powell singular                      4      17      17     2.945101e-09   1.660249e-02
Powell singular-AD                   4      17      17     2.945101e-09   9.537205e-03
Powell badly scaled                  2      15      14     7.051804e-12   1.120495e-02
Powell badly scaled-AD               2      15      14     7.051804e-12   1.071290e-02
Wood                                 4      19      15     1.152565e-10   2.748516e-02
Wood-AD                              4      19      15     1.152340e-10   1.655836e-02
Helical Valley                       3      10       9     1.110223e-14   1.909693e-02
Helical Valley-AD                    3      10       9     1.110223e-14   4.237322e-02
Watson                               6      20      16     2.848399e-13   5.379833e-02
Watson-AD                            6      20      16     2.783641e-13   4.923553e-02
Watson                               9      17      17     1.591530e-12   2.750510e-04
Watson-AD                            9      17      17     1.623697e-12   1.462754e-02
Chebyquad                            5       8       5     9.143861e-09   2.286787e-02
Chebyquad-AD                         5       8       5     9.143861e-09   2.172272e-02
Chebyquad                            6      13       8     7.320533e-16   7.294300e-05
Chebyquad-AD                         6      13       8     7.320533e-16   1.644340e-04
Chebyquad                            7      10       6     6.329061e-10   4.751800e-05
Chebyquad-AD                         7      10       6     6.329061e-10   1.759580e-04
Chebyquad                            9      14       8     1.616762e-14   7.264700e-05
Chebyquad-AD                         9      14       8     1.616762e-14   5.661960e-04
Brown almost-linear                 10       5       5     3.382331e-09   6.716922e-02
Brown almost-linear-AD              10       5       5     3.382331e-09   6.762144e-02
Brown almost-linear                 30       4       4     8.883262e-09   1.450521e-03
Brown almost-linear-AD              30       4       4     8.883262e-09   1.454770e-04
Brown almost-linear                 40       4       4     1.570836e-09   1.464160e-04
Brown almost-linear-AD              40       4       4     1.570836e-09   1.757550e-04
Discrete boundary value             10       4       4     2.636780e-16   2.289637e-02
Discrete boundary value-AD          10       4       4     2.636780e-16   1.592200e-02
Discrete integral equation           1       4       4     8.540391e-14   2.805541e-02
Discrete integral equation-AD        1       4       4     8.540391e-14   2.447792e-02
Discrete integral equation          10       4       4     2.720046e-15   1.865940e-04
Discrete integral equation-AD       10       4       4     2.720046e-15   7.060400e-04
Trigonometric                       10      14       8     1.204317e-10   1.970853e-02
Trigonometric-AD                    10      14       8     1.204299e-10   1.603536e-02
Variably dimensioned                10      15      15     1.323497e-12   2.271842e-02
Variably dimensioned-AD             10      15      15     1.323497e-12   1.395246e-02
Broyden tridiagonal                 10       5       5     7.548402e-10   1.479347e-02
Broyden tridiagonal-AD              10       5       5     7.548402e-10   1.319243e-02
Broyden banded                      10       6       6     9.359466e-09   2.366735e-02
Broyden banded-AD                   10       6       6     9.359466e-09   1.647761e-02
	PASSED: minpack.jl
	PASSED: iface.jl
	PASSED: already_converged.jl
	PASSED: autodiff.jl
	PASSED: josephy.jl
	PASSED: difficult_mcp.jl
	PASSED: sparse.jl
```

I also compared timings: 

```
julia> extrema(master - pr)
(-0.007460460000000002,0.006471019999999994)

julia> mean(master - pr)
2.3107523809524104e-5
```

Doesn't seem to make a huge impact, but I like the idea.

**Comment by spencerlyon2** [comment]
Thanks!

**Comment by matthieugomez** [comment]
Thanks for merging! You can see an example of improved timing in BansalYaronProblemPowell (that solves a non linear finite difference scheme using NLsolve) in https://github.com/matthieugomez/HJBFiniteDifference.jl

**Comment by spencerlyon2** [comment]
cool, thanks for sharing.

**Pull request merged**